### PR TITLE
Mirgecompare: check that rank numbers match

### PR DESCRIPTION
**Pull request by poulNA**
Rank number feature by XML parsing through pvtu files
**Questions for the review**:
- [ ] Is the scope and purpose of the PR clear?
  - [ ] The PR should have a description.
  - [ ] The PR should have a guide if needed (e.g., an ordering).
- [ ] Is every top-level method and class documented? Are things that should be documented actually so?
- [ ] Is the interface understandable? (I.e. can someone figure out what stuff does?) Is it well-defined?
- [ ] Does the implementation do what the docstring claims?
- [ ] Is everything that is implemented covered by tests?
- [ ] Do you see any immediate risks or performance disadvantages with the design? Example: what do interface normals attach to?

### CODE REVIEW

**Comment by poulNA** [comment]
### Example Output

**Comparing file from a single rank and a file from a double rank** 
Now warns if compared files were run with different ranks. Otherwise no warning

Command: `bin/mirgecompare.py 1rank/fld-wave-mpi-000-0000.pvtu 2rank/fld-wave-mpi-000-0000.pvtu`

Output: 
```
File 1rank/fld-wave-mpi-000-0000.pvtu has 1 rank, but File 2rank/fld-wave-mpi-000-0000.pvtu has 2 rank.
  warn(f"File {first_file} has {rank1} rank, "
Field: u Max Error: 0.96754678098049
Field: v Max Error: 0
Traceback (most recent call last):
  File "/home/paulj3/emirge/mirgecom/bin/mirgecompare.py", line 37, in <module>
    compare_files_vtu(first_file, second_file, file_type, user_tolerance)
  File "/home/paulj3/emirge/mirgecom/mirgecom/simutil.py", line 533, in compare_files_vtu
    raise ValueError("Fidelity test failed: Mismatched data array "
ValueError: Fidelity test failed: Mismatched data array values tolerance=1e-12.
```

**Comment by MTCam** [comment]
> ### Example Output
> **Comparing file from a single rank and a file from a double rank** Now warns if compared files were run with different ranks. Otherwise no warning

Hi @poulNA and @matthiasdiener.  I am a little confused why it should be a warning and not just an error.  @matthiasdiener said it is a warning because you guys have to parse the XML yourselves to determine `nranks`.   Why does that preclude an error?

**Comment by poulNA** [comment]
> > ### Example Output
> > **Comparing file from a single rank and a file from a double rank** Now warns if compared files were run with different ranks. Otherwise no warning
> 
> Hi @poulNA and @matthiasdiener. I am a little confused why it should be a warning and not just an error. @matthiasdiener said it is a warning because you guys have to parse the XML yourselves to determine `nranks`. Why does that preclude an error?

I believe it’s because we decided not to halt the execution in the case of different nranks because there might be some useful information. And the code executes fully without running into any other major problems.  Hopefully @matthiasdiener can provide more reason.

**Comment by MTCam** [comment]
> > Hi @poulNA and @matthiasdiener. I am a little confused why it should be a warning and not just an error. @matthiasdiener said it is a warning because you guys have to parse the XML yourselves to determine `nranks`. Why does that preclude an error?
> 
> I believe it’s because we decided not to halt the execution in the case of different nranks because there might be some useful information. And the code executes fully without running into any other major problems. Hopefully @matthiasdiener can provide more reason.

Comparing two VTU sets run on different number of ranks has no value unless we can stitch the meshes and data together from both sets back into a fully assembled mesh.  There is just no way to correlate the VTU files.   Trying to do it should result in an error, imo.

**Comment by matthiasdiener** [comment]
> > ### Example Output
> > **Comparing file from a single rank and a file from a double rank** Now warns if compared files were run with different ranks. Otherwise no warning
> 
> Hi @poulNA and @matthiasdiener. I am a little confused why it should be a warning and not just an error. @matthiasdiener said it is a warning because you guys have to parse the XML yourselves to determine `nranks`. Why does that preclude an error?

I wasn't sure whether there are other circumstances in which a pvtu file could have multiple pieces, e.g. due to exceeding some file size limits or so. If we are reasonably sure that there aren't such circumstances, we might as well convert this to an error.

File: mirgecom/simutil.py [hunk 1]
```diff
@@ -511,6 +512,19 @@ def compare_files_vtu(
...
+    def numranks(filename):
+        tree = Et.parse(filename)
+        root = tree.getroot()
+        return len(root.findall(".//Piece"))
+
+    if file_type == "pvtu":
+        rank1 = numranks(first_file)
+        rank2 = numranks(second_file)
+        if rank1 is not rank2:
+            warn(f"File {first_file} has {rank1} rank, "
```
Comment (matthiasdiener) [inline mirgecom/simutil.py:525]:
```suggestion
            warn(f"File '{first_file}' has {rank1} ranks, "
```

**Comment by MTCam** [comment]
> I wasn't sure whether there are other circumstances in which a pvtu file could have multiple pieces, e.g. due to exceeding some file size limits or so. If we are reasonably sure that there aren't such circumstances, we might as well convert this to an error.

afaik the only way to get multiple VTU files at a single time level in a single run is to write a partitioned dataset, and there is also not a built-in way through VTK to assemble I think.    If we are ever to be able to compare two runs of differing Nranks (which we do really need!), we'll need to assemble both datasets back to their serial representation (ugh).  If we switch to HDF5, or another format which can do single file (serializes on-the-fly), then we would not have this limitation.  

This basic problem is just true for formats and i/o strategies where we write one file per partition (rank).

**Comment by poulNA** [comment]
> > I wasn't sure whether there are other circumstances in which a pvtu file could have multiple pieces, e.g. due to exceeding some file size limits or so. If we are reasonably sure that there aren't such circumstances, we might as well convert this to an error.
> 
> afaik the only way to get multiple VTU files at a single time level in a single run is to write a partitioned dataset, and there is also not a built-in way through VTK to assemble I think. If we are ever to be able to compare two runs of differing Nranks (which we do really need!), we'll need to assemble both datasets back to their serial representation (ugh). If we switch to HDF5, or another format which can do single file (serializes on-the-fly), then we would not have this limitation.
> 
> This basic problem is just true for formats and i/o strategies where we write one file per partition (rank).

Should I then go ahead and change the warning to an error?

**Comment by matthiasdiener** [comment]
> > > I wasn't sure whether there are other circumstances in which a pvtu file could have multiple pieces, e.g. due to exceeding some file size limits or so. If we are reasonably sure that there aren't such circumstances, we might as well convert this to an error.
> > 
> > 
> > afaik the only way to get multiple VTU files at a single time level in a single run is to write a partitioned dataset, and there is also not a built-in way through VTK to assemble I think. If we are ever to be able to compare two runs of differing Nranks (which we do really need!), we'll need to assemble both datasets back to their serial representation (ugh). If we switch to HDF5, or another format which can do single file (serializes on-the-fly), then we would not have this limitation.
> > This basic problem is just true for formats and i/o strategies where we write one file per partition (rank).
> 
> Should I then go ahead and change the warning to an error?

Yes please.

File: mirgecom/simutil.py [hunk 2]
```diff
@@ -511,6 +511,19 @@ def compare_files_vtu(
...
+        tree = Et.parse(filename)
+        root = tree.getroot()
+        return len(root.findall(".//Piece"))
+
+    if file_type == "pvtu":
+        rank1 = numranks(first_file)
+        rank2 = numranks(second_file)
+        if rank1 is not rank2:
+            raise ValueError(f"File '{first_file}' has {rank1} ranks, "
+                f"but File {second_file} has {rank2} ranks.")
```
Comment (matthiasdiener) [inline mirgecom/simutil.py:525]:
```suggestion
                f"but file ‘{second_file}’ has {rank2} ranks.")
```

File: mirgecom/simutil.py [hunk 3]
```diff
@@ -511,6 +511,19 @@ def compare_files_vtu(
     reader1.Update()
     output1 = reader1.GetOutput()
 
+    # Check rank number
+    def numranks(filename):
```
Comment (matthiasdiener) [inline mirgecom/simutil.py:515]:
```suggestion
    def numranks(filename: str) -> int:
```

File: mirgecom/simutil.py [hunk 4]
```diff
@@ -511,6 +511,19 @@ def compare_files_vtu(
...
+        tree = Et.parse(filename)
+        root = tree.getroot()
+        return len(root.findall(".//Piece"))
+
+    if file_type == "pvtu":
+        rank1 = numranks(first_file)
+        rank2 = numranks(second_file)
+        if rank1 is not rank2:
```
Comment (matthiasdiener) [inline mirgecom/simutil.py:523]:
```suggestion
        if rank1 != rank2:
```

**Review by matthiasdiener — COMMENTED** [review]
Apart from these minor comments, LGTM!

**Review by matthiasdiener — APPROVED** [review]
LGTM, thanks!

**Review by MTCam — APPROVED** [review]
This looks great!  Thanks, Paul!

**Pull request merged**